// phovea_ui/src/buildInfo.js
/**
 * Created by Samuel Gratzl on 21.11.2016.
 */
var BuildInfo = (function () {
    function BuildInfo(buildInfo) {
        this.buildInfo = buildInfo;
    }
    BuildInfo.prototype.toString = function () {
        return 'BuildInfo';
    };
    BuildInfo.prototype.buildBuildInfo = function () {
        var build = this.buildInfo;
        return "<table class=\"table table-bordered table-condensed\">\n            <tbody>\n              <tr><th>Application</th><td>" + build.name + "</td></tr>\n              <tr><th>Version</th><td>" + build.version + "</td></tr>\n              <tr><th>Url</th><td><code>" + location.pathname + location.hash + "</code></td></tr>\n              <tr><th>UserAgent</th><td>" + navigator.userAgent + "</td></tr>\n              <tr><th>GitHub</th><td><a href=\"" + build.resolved.replace(/\.git.*/, '/issues/new') + "\" target=\"_blank\">Submit Issue (include the build details below)</a></td></tr>\n            </tbody>\n            </table>";
    };
    BuildInfo.prototype.buildIssuesContent = function () {
        // don't use location.href, since I don't wanna expose server names
        return "\nKey | Value\n--- | -----\nApplication | " + this.buildInfo.name + "\nVersion | [" + this.buildInfo.version + "](" + this.buildInfo.resolved.replace('.git#', '/commit/') + ")\nUrl | `" + location.pathname + location.hash + "`\nUserAgent | " + navigator.userAgent + "\nPlatform | " + navigator.platform + "\nScreen Size | " + screen.width + " x " + screen.height + "\nWindow Size | " + window.innerWidth + " x " + window.innerHeight + "\n      \n~~~json\n" + JSON.stringify(this.buildInfo, null, ' ') + "\n~~~";
    };
    BuildInfo.prototype.toHTML = function () {
        return "\n      <h4>Build Info</h4>\n      " + this.buildBuildInfo() + "\n      <h4>Build Details</h4>\n      <textarea readonly=\"readonly\">" + this.buildIssuesContent() + "</textarea>\n    ";
    };
    return BuildInfo;
}());
export { BuildInfo };
export default function build() {
    return self.fetch('./buildInfo.json').then(function (response) { return response.json(); }).then(function (result) { return new BuildInfo(result); });
}
