// phovea_importer/src/parser.js
/**
 * Created by Samuel Gratzl on 29.09.2016.
 */
import { mixin } from 'phovea_core/src/index';
import * as papaparse from 'papaparse';
var defaultOptions = {
    skipEmptyLines: true
};
/**
 * parses the given csv file/blob using PapaParse
 * @param data
 * @param options additional options
 * @return {Promise<R>|Promise}
 */
export function parseCSV(data, options) {
    if (options === void 0) { options = {}; }
    return new Promise(function (resolve, reject) {
        papaparse.parse(data, mixin({
            complete: function (result) { return resolve({ data: result.data, meta: result.meta }); },
            error: function (error) { return reject(error); }
        }, defaultOptions, options));
    });
}
export function streamCSV(data, chunk, options) {
    if (options === void 0) { options = {}; }
    return new Promise(function (resolve, reject) {
        papaparse.parse(data, mixin({
            complete: function (result) { return resolve(null); },
            chunk: function (result) { return chunk({ data: result.data, meta: result.meta }); },
            error: function (error) { return reject(error); }
        }, defaultOptions, options));
    });
}
